pybind11/setup_helpers.py

```diff
@@ -0,0 +1,276 @@
+# -*- coding: utf-8 -*-
+
+"""
+This module provides helpers for C++11+ projects using pybind11.
+
+LICENSE:
+
+Copyright (c) 2016 Wenzel Jakob <wenzel.jakob@epfl.ch>, All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice, this
+   list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+
+import contextlib
+import os
+import shutil
+import sys
+import tempfile
+import threading
+import warnings
+
+try:
+    from setuptools.command.build_ext import build_ext as _build_ext
+    from setuptools import Extension as _Extension
+except ImportError:
+    from distutils.command.build_ext import build_ext as _build_ext
+    from distutils.extension import Extension as _Extension
```

Comment:
I would like to try to make this dual-compatible (basically to support #2444, as you can't ensure setuptools is installed, though you can ensure it is installed when running setup.py).

Reply:
(also, other than this one set of lines, pretty much nothing changes)

Reply:
(setuptools wraps or vendors and wraps distutils, depending on the setuptools version, and you normally get setuptools wrapped libs from distutils anyway as long as you import setuptools first or run from pip, so they basically have to have about the same API)

Reply:
OK, fine, then :-) Just a suggestion, if things become too incompatible (or if `setuptools` would e.g. have `has_flag` or some tools like that you'd like to reuse).

Reply:
A heads up, they are planning on deprecating and removing distutils ([PEP 632](https://www.python.org/dev/peps/pep-0632/)). I'm not sure if the version in setuptools is importable as distutils or it's for internal use only. I would probably try to future proof this.

Reply:
This is only a fallback if setuptools is not available, so it should be safe. If, in Python 3.11 or 3.12 distutils disappears, then this still imports setuptools before trying distutils, and if you have neither, it's not going to work (but it would never have worked)

Reply:
For the actual setup.py, we simply require setuptools. This is just to help support JIT type compilation too, where setuptools is not as likely to be mandated (I think it's pretty commonly added to venvs/conda envs automatically? PEP 518 makes it no longer absolutely required, though, so even then, it might not continue to be available)